File: sample_app_dagger/src/main/java/com/chumarin/stanislav/sample_app_dagger/login/LoginViewModel.java
package com.chumarin.stanislav.sample_app_dagger.login;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.chumarin.stanislav.sample_app_dagger.Session;
import com.chumarin.stanislav.sample_app_dagger.util.BaseViewModel;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.BehaviorSubject;
import vit.khudenko.android.sessiontracker.SessionRecord;
import vit.khudenko.android.sessiontracker.SessionTracker;

import javax.inject.Inject;
import java.util.Optional;
import java.util.concurrent.TimeUnit;

public class LoginViewModel extends BaseViewModel {

    @NonNull
    private final SessionTracker<Session.Event, Session.State> sessionTracker;
    @NonNull
    private final BehaviorSubject<State> state = BehaviorSubject.createDefault(State.Idle.INSTANCE);

    @Nullable
    private Disposable disposable;

    @Inject
    LoginViewModel(@NonNull SessionTracker<Session.Event, Session.State> sessionTracker) {
        this.sessionTracker = sessionTracker;
    }

    void onLoginButtonClicked(@NonNull String userId) {
        state.onNext(State.Progress.INSTANCE);

        disposable = Observable.fromCallable(() -> {
            synchronized (sessionTracker) {
                Optional<SessionRecord<Session.State>> optionalSessionRecord = sessionTracker.getSessionRecords()
                        .stream()
                        .filter(record -> record.sessionId().equals(userId))
                        .findFirst();
                if (optionalSessionRecord.isPresent()) {
                    sessionTracker.consumeEvent(userId, Session.Event.LOGIN);
                } else {
                    sessionTracker.trackSession(userId, Session.State.ACTIVE);
                }
            }
            return userId;
        })
                .subscribeOn(Schedulers.io())
                .delay(2, TimeUnit.SECONDS, Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(loggedInUserId -> state.onNext(new State.Success(loggedInUserId)));
    }

    Observable<State> observeState() {
        return state.hide();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        if (disposable != null) {
            disposable.dispose();
            disposable = null;
        }
    }

    static abstract class State {
        private State() {
        }

        static class Idle extends State {
            static final Idle INSTANCE = new Idle();

            private Idle() {
            }
        }

        static class Progress extends State {
            static final Progress INSTANCE = new Progress();

            private Progress() {
            }
        }

        static class Success extends State {
            @NonNull
            private final String userId;

            Success(@NonNull String userId) {
                this.userId = userId;
            }

            @NonNull
            String getUserId() {
                return userId;
            }
        }
    }
}

File: sample_app_koin/src/main/kotlin/vit/khudenko/android/sessiontracker/sample/koin/login/LoginViewModel.kt
package vit.khudenko.android.sessiontracker.sample.koin.login

import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import vit.khudenko.android.sessiontracker.SessionId
import vit.khudenko.android.sessiontracker.SessionTracker
import vit.khudenko.android.sessiontracker.sample.koin.Session
import vit.khudenko.android.sessiontracker.sample.koin.util.BaseViewModel

class LoginViewModel(
    private val sessionTracker: SessionTracker<Session.Event, Session.State>
) : BaseViewModel() {

    private val state = MutableStateFlow<State>(State.Idle)

    fun onLoginButtonClicked(sessionId: SessionId) {
        state.value = State.Progress

        viewModelScope.launch {
            val targetSessionRecord = sessionTracker.getSessionRecords()
                .firstOrNull { record -> record.sessionId == sessionId }
            if (targetSessionRecord == null) {
                sessionTracker.trackSession(sessionId, Session.State.ACTIVE)
            } else {
                sessionTracker.consumeEvent(sessionId, Session.Event.LOGIN)
            }
            delay(2000)
            state.value = State.Success(sessionId)
        }
    }

    fun stateFlow(): Flow<State> = state.asStateFlow()

    sealed class State {
        object Idle : State()
        object Progress : State()
        class Success(val sessionId: SessionId) : State()
    }
}
